lib/location/method_wifi.c

```diff
@@ -0,0 +1,307 @@
+
+	if (current_scan_result_count <= CONFIG_LOCATION_METHOD_WIFI_SCANNING_RESULTS_MAX_CNT) {
+		sprintf(current->mac_addr_str,
+			"%02x:%02x:%02x:%02x:%02x:%02x",
+			entry->mac[0], entry->mac[1], entry->mac[2],
+			entry->mac[3], entry->mac[4], entry->mac[5]);
+		snprintf(current->ssid_str, entry->ssid_length + 1, "%s", entry->ssid);
+
+		current->channel = entry->channel;
+		current->rssi = entry->rssi;
+
+		LOG_DBG("scan result #%d stored: ssid %s, mac address: %s, channel %d,",
+			current_scan_result_count,
+			log_strdup(current->ssid_str),
+			log_strdup(current->mac_addr_str),
+			current->channel);
+	} else {
+		LOG_WRN("Scanning result (mac %02x:%02x:%02x:%02x:%02x:%02x) "
+			"did not fit to result buffer - dropping it",
+				entry->mac[0], entry->mac[1], entry->mac[2],
+				entry->mac[3], entry->mac[4], entry->mac[5]);
+	}
+}
+
+static void method_wifi_handle_wifi_scan_done(struct net_mgmt_event_callback *cb)
+{
+	const struct wifi_status *status = (const struct wifi_status *)cb->info;
+
+	if (status->status) {
+		LOG_WRN("WiFi scan request failed (%d).", status->status);
+	} else {
+		LOG_INF("Scan request done.");
+	}
+	k_sem_give(&wifi_scanning_ready);
+
+	latest_scan_result_count =
+		(current_scan_result_count > CONFIG_LOCATION_METHOD_WIFI_SCANNING_RESULTS_MAX_CNT) ?
+			CONFIG_LOCATION_METHOD_WIFI_SCANNING_RESULTS_MAX_CNT :
+			current_scan_result_count;
+	current_scan_result_count = 0;
+}
+
+static struct net_mgmt_event_callback method_wifi_net_mgmt_cb;
+
+void method_wifi_net_mgmt_event_handler(struct net_mgmt_event_callback *cb, uint32_t mgmt_event,
+					struct net_if *iface)
+{
+	if (running) {
+		switch (mgmt_event) {
+		case NET_EVENT_WIFI_SCAN_RESULT:
+			method_wifi_handle_wifi_scan_result(cb);
+			break;
+		case NET_EVENT_WIFI_SCAN_DONE:
+			method_wifi_handle_wifi_scan_done(cb);
+			break;
+		default:
+			break;
+		}
+	}
+}
+
+/******************************************************************************/
+
+static void method_wifi_positioning_work_fn(struct k_work *work)
+{
+	struct loc_location location_result = { 0 };
+	struct k_work_args *work_data = CONTAINER_OF(work, struct k_work_args, work_item);
+	const struct loc_wifi_config wifi_config = work_data->wifi_config;
+	int64_t starting_uptime_ms = work_data->starting_uptime_ms;
+	int err;
+
+	loc_core_timer_start(wifi_config.timeout);
+
+	err = method_wifi_scanning_start();
+	if (err) {
+		LOG_WRN("Cannot start WiFi scanning, err %d", err);
+		goto return_error;
+	}
+
+	k_sem_take(&wifi_scanning_ready, K_FOREVER);
+	if (running) {
+		struct rest_wifi_pos_request request;
+		struct rest_wifi_pos_result result;
+
+		if (!loc_utils_is_default_pdn_active()) {
+			/* No worth to start trying to fetch with the REST api over cellular.
+			 * Thus, fail faster in this case and safe the trying "costs".
+			 */
+			LOG_WRN("Default PDN context is NOT active, cannot retrieve a location");
+			err = -1;
+			goto return_error;
+		}
```

Comment:
I'm wondering whether this is necessary or not. Yes, it does save unnecessary work if the PDN is not active, but then again that's most likely a programming error in the application. Or I guess when device is out of service, although also in that case the default PDN most likely remains active for some time when modem is trying to find a new cell. Scanning has also already been done and the energy used. Most likely this would just fail pretty quickly in some socket operation?

Reply:
I'd rather keep this as it is. 
Clear cause why it is failing, socket operation might fail (later) but the cause in there is not clear.
Temporarily out of service case is ok to let to go to more time consuming socket level.